### DISCUSSION TITLE
Protected/Privileged References

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#7021] [Language Ideas]
This arose out of a sidebar around extensions regarding their access to the members of the underlying type:  https://github.com/dotnet/csharplang/issues/5497#issuecomment-1451027006

There are use cases where a developer might want to write a helper/extension method to be used specifically within a derived type.  The problem is that any such helper would necessarily be declared in a separate type outside of the type hierarchy and would not have access to any protected members.

```mermaid
graph LR
    subgraph Third-party classes
        A --> B
        A --> C
    end
    subgraph Classes in solution
        B --> D
        C --> E
        C --> F
        D -.->|calls| Helper[Helper calling protected A methods]
        E -.->|calls| Helper
        F -.->|calls| Helper
    end
```
_Diagram credit to @jnm2_ 

```cs
public class FrameworkElement {
    protected internal DependencyObject GetTemplateChild(string childName);
}

public static class FrameworkElementHelpers {
    public static T? GetTemplateChild<T>(this FrameworkElement element, string childName) where T : DependencyObject {
        // error CS0122: 'FrameworkElement.GetTemplateChild(string)' is inaccessible due to its protection level
        DependencyObject child = element.GetTemplateChild(childName);

        if (child is T typed) {
            return typed;
        }

        // other stuff
        return null;
    }
}
```

For cases where the developer can't write a common derived type this forces developers to effectively copy and paste the logic between multiple types, or to use abstractions to pass the protected members to the helper class.

I'd like to propose a mechanism to take and pass "privileged" references.  Those references are granted access to the protected members of the type:

```cs
// given
public class Foo {
    protected void ProtectedMethod() { }
}

// the developer could write the following helper
public static class FooHelpers {
    public static void HelperMethod(protected Foo instance) {
        // allowed
        instance.ProtectedMethod();
    }

    public static void ExtensionMethod(protected this Foo instance) {
        // allowed
        instance.ProtectedMethod();
    }
}
```

When passing a parameter of a privileged reference the compiler enforces that the reference can only be passed from a context in which the developer may already call protected members on that reference:

```cs
public class Bar : Foo {
    public void Test() {
        // allowed
        FooHelpers.HelperMethod(this);

        // also allowed
        this.HelperMethod();
    }
}

public class OtherType {
    public void Test(Foo foo) {
        // illegal, cannot access protected members of Foo here
        FooHelpers.HelperMethod(foo);
    }
}
```

**Drawbacks**
* Complicates accessibility
* Requires runtime changes to relax member access protected

**Alternatives**
* Could be made to apply to extensions only:
    ```cs
    public extension FooExtensions for protected Foo {
        // ...
    }
    ```

**Workaround**
* It is currently possible to write these kinds of helper methods, but only if the protected members are passed as delegates.
* Source generators can be used to copy/paste the same logic between the different derived types.

**Comment by iam3yal** [comment]
Another workaround that some developer use today is through namespaces where the name is something like `.Internal`, `.Protected` or something similar.

**Reply by HaloFour** [reply]
That only works if you are the author of the base class, which is often not the case.

**Reply by iam3yal** [reply]
@HaloFour What do you mean? in the case I'm talking about the methods are `public` not `protected` but then this `ProtectedMethod` needs to move to another type.

**Reply by HaloFour** [reply]
@eyalalonn 

I'm not sure what you're suggesting.  If you do not own the assembly in which the base type is defined (e.g. `FrameworkElement`) then you can't change the accessibility of the method or move it anywhere else anyway.  If those members on that type are `protected` then your options are pretty limited.

**Reply by iam3yal** [reply]
@HaloFour I'm not suggesting anything 😄 I'm saying that there are alternatives today but they aren't great and I think that your proposed solution is pretty good to have.

> If you do not own the assembly in which the base type is defined (e.g. FrameworkElement) then you can't change the accessibility of the method or move it anywhere else anyway. If those members on that type are protected then your options are pretty limited.

The reason I'm saying that some developers by design workaround it but avoiding having `internal` and `protected` members and instead use `public` and then put them into "special" namespaces that 3rd party can use to access "core" functionality.

**Reply by michael-hawker** [reply]
Yeah, the main scenario for this is writing helpers for using other library's code where you don't have access to change the underlying infrastructure. If that was the case, there'd be less problems as you could change or implement the helpers you needed in the base class. 🙂

**Reply by timcassell** [reply]
> Requires runtime changes to relax member access protected

Perhaps the compiler could emit `IgnoresAccessChecksToAttribute` in lieu of a runtime change?

**Reply by svick** [reply]
@timcassell I think that would require for that attribute to become officially supported. No idea how likely that is.

**Reply by HaloFour** [reply]
@timcassell 

That looks like it works the other way around, like IVT, no?

**Reply by timcassell** [reply]
@HaloFour No, it's the inverse of that. Someone even tried to add support for it, but it was rejected. https://github.com/dotnet/roslyn/pull/20870 The compiler should be able to use it itself internally without language support.

**Reply by HaloFour** [reply]
@timcassell 

Sounds like that might be part of the runtime mechanism that can be used to support these scenarios.  I'd rather see the language support this scenario through specific code and patterns, though, where proper encapsulation can continue to be enforced.  I can't imagine that the team would be that willing to honor an attribute that otherwise completely disables accessibility.

**Comment by DavidArno** [comment]
I really don't like this idea. The purpose of extension methods is to add subsidiary functionality to types and they are by-design deliberately restricted to only accessing public members.

Rather than copying and pasting the logic between multiple types, write a source generator that applies that logic to those multiple types. This works as-is and avoids runtime changes and migrating away from the current principles underlying extension methods.

**Reply by jnm2** [reply]
The motivation in my cases is code deduplication to make code maintenance less of a burden for uniform behaviors across all classes inheriting from the third-party classes. Writing a bespoke source generator for each situation is not a way to reduce the burden of maintaining this code.

**Reply by HaloFour** [reply]
@DavidArno

I'll add source generators to the list of workarounds.

**Reply by michael-hawker** [reply]
Writing sources generators is non-trivial compared to having a simple extension method in a known pattern. While source generators are great, they're not accessible as a tool for every scenario.

That would also still duplicate the underlying code and increase the binary size compared to having a central extension method which can be reused by any callers.

**Reply by DavidArno** [reply]
@michael-hawker,

> Writing sources generators is non-trivial compared to having a simple extension method in a known pattern...

Changing the runtime, changing the C# language and changing all the documentation across the whole of the web to allow for extension methods to access protected members is very much non-trivial. In fact, unless there are many thousands of developers needing to do this a great many times, I'd argue that writing a few source generators to handle this is vanishingly trivial by comparison.

But it sounds like the runtime may be coming to your rescue here from what @svick says below. A zero-overhead bypassing of acess protection of non-public members is what is being asked for here, and it sounds like that's what .NET 8 may well deliver. So everyone can be happy...

**Comment by svick** [comment]
In .Net 8, you will (likely) be able to bypass access checks using `extern` methods with zero overhead, see https://github.com/dotnet/runtime/issues/81741. So another approach would be to generate methods to access the `protected` members. 

But I don't know how you would generate something that would actually work well. (If you generate extension methods, you can't easily limit their scope; if you generate them as `private`, you can't invoke them as instance methods.)
